### ISSUE TITLE
Crash during D3D11 capture when a buffer is mapped simultaneously on 2 threads

### ISSUE THREAD

**Issue by arkgmarion** [baldurk/renderdoc#1913]
## Description

During a D3D11 capture, a race condition can happen if the same buffer is mapped at exactly the same time on 2 different deferred contexts, for the first time.

## Steps to reproduce

Quite difficult to reproduce as it is a multi-threading bug

1. initialize a buffer and 2 D3D11 deferred contexts
2. spawn 2 threads
3. those 2 threads call Map on the buffer, each on one context
4. during the capture of step 3, observe a crash in WrappedID3D11DeviceContext::Serialise_Map, often line 7467 :       memcpy(record->GetShadowPtr(ctxMapID, 1), appMem, mapLength);

This is due to AllocShadowStorage & GetShadowPtr not being thread-safe.
Different problems can happen : 
- the 2 threads can end up calling AllocShadowStorage on the same record
- one thread could be in the middle of AllocShadowStorage when the other one calls GetShadowPtr which can lead to ptr[0] being allocated but ptr[1] still being NULL

I fixed it by adding SCOPED_LOCK(DeferredShadowLock); in those 2 functions but I'm not sure it is the best solution as it is the first time I ever touch RenderDoc's code.

## Environment

* RenderDoc version: 1.6
* Operating System: Windows 10 1909
* Graphics API: D3D11

**Comment by baldurk** [comment]
I think I see the problem, but just to verify my understanding against what you're seeing: the problem looks like it'd be particularly with `GetContextID()` which resizes the deferred shadow array. If that happens on one thread while another thread is calling `AllocShadowStorage` that would race and you could end up with a thread which allocated in the old destroyed array and its pointer didn't get copied across.

The intention is that once `DeferredShadow` has a slot allocated for the context then it can call AllocShadowStorage/GetShadowPtr freely on its own slot without needing lock, since other threads access other slots.

ie. imagine that the array was fixed size, or got `reserve()`'d up to a big enough number that it didn't resize, I don't believe there would be a race since there's a lock covering each thread getting its slot and from then on each deferred context accesses its own data.

The solution I suspect is similar to the lock you added, though it'll probably be better to make this a read-write lock so that threads can freely multithread on the array with a read-lock unless it's being resized which would take a write-lock.

**Comment by arkgmarion** [comment]
You’re right, I missed the ContextID part. The realloc of the DeferredShadow array while another thread is accessing it (either to alloc or get the ptr members) is probably the root cause of the crash.

I had to guess what the real problem could be because by the time the game crashes, both threads have already moved forward in their execution and have returned from GetContextID, AllocShadowStorage and GetShadowPtr.

I’m not very familiar with RW Locks advantages over regular mutexes, it probably depends on their implementation.

I’ll leave the final fix choice to you as RenderDoc is your baby (and a beautiful one that we use every day)

By the way, I have experienced another crash around the buffer code but I have seen it only once and don’t have much info to share yet except that it was during the Unmap call.
I’ve not managed to reproduced it again, it seems to be much rarer than the Map crash. I’ll let you know if I get it again in the next few days.

Thanks,
Gilles

**Comment by baldurk** [comment]
Thanks for clarifying, sounds like we're on the same page. The main advantage of the RW lock is that multiple readers can hold the lock simultaneously, it's only if you need to take a write lock that it actually prevents other threads from entering. That works nicely for this situation where it will prevent threads from serialising once slots have been allocated.

**Comment by baldurk** [comment]
I was able to reasonably reliably reproduce this by setting up a demo to create a buffer then immediately send a bunch of threads mapping it. That commit fixed it for me, please let me know if it doesn't fix your issue and I can look into it further.

**Comment by arkgmarion** [comment]
I have not been able to repro the crash since I integrated the latest renderdoc github head revision, so I guess it is fixed, thanks.

However, after this upgrade (from 1.6) I experienced another completely unrelated systematic crash during renderdoc initialization that I had to fix before being able to test, I'm opening another issue right now to discuss it.
